{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# STATS 604: Gene expression in the Brain\n",
    "## Team Members: Ben Agyare, Yumeng Wang, Jake Trauger, Yash Patel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code Setup\n",
    "We start simply by loading in the data and organizing properties of subsequent interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "with open(\"data/brain.pkl\", \"rb\") as f:\n",
    "    dataset = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression = dataset[\"expression\"]\n",
    "genes = dataset[\"genes\"]\n",
    "samples = dataset[\"samples\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Normalization\n",
    "We need to do some normalization with the bacterial spike-ins. According to the following two sources:\n",
    "- https://support.bioconductor.org/p/49150/\n",
    "- https://bioinformatics.mdanderson.org/MicroarrayCourse/Lectures/ma07b.pdf\n",
    "two common normalization techniques are spike-in normalizatio and quantile normalization. The former doesn't really involve much assumption beyond the experimental consistency of spike-in amounts. Given that this itself could vary across labs, we should verify this behavior prior to applying spike-in normalization. Quantile normalization assumes that most genes are *not* differentially expressed (that only a handful are). \n",
    "\n",
    "Let's start with the simple technique."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AFFX-BioB-3_at', 'AFFX-BioB-3_st', 'AFFX-BioB-5_at', 'AFFX-BioB-5_st', 'AFFX-BioB-M_at', 'AFFX-BioB-M_st', 'AFFX-BioC-3_at', 'AFFX-BioC-3_st', 'AFFX-BioC-5_at', 'AFFX-BioC-5_st', 'AFFX-BioDn-3_st', 'AFFX-BioDn-5_at', 'AFFX-BioDn-5_st', 'AFFX-CreX-3_at', 'AFFX-CreX-3_st', 'AFFX-CreX-5_at', 'AFFX-CreX-5_st', 'AFFX-DapX-3_at', 'AFFX-DapX-5_at', 'AFFX-DapX-M_at', 'AFFX-HUMRGE/M10098_3_at', 'AFFX-HUMRGE/M10098_5_at', 'AFFX-HUMRGE/M10098_M_at', 'AFFX-LysX-3_at', 'AFFX-LysX-5_at', 'AFFX-LysX-M_at', 'AFFX-M27830_3_at', 'AFFX-M27830_5_at', 'AFFX-M27830_M_at', 'AFFX-MurFAS_at', 'AFFX-MurIL10_at', 'AFFX-MurIL2_at', 'AFFX-MurIL4_at', 'AFFX-PheX-3_at', 'AFFX-PheX-5_at', 'AFFX-PheX-M_at', 'AFFX-ThrX-3_at', 'AFFX-ThrX-5_at', 'AFFX-ThrX-M_at', 'AFFX-TrpnX-3_at', 'AFFX-TrpnX-5_at', 'AFFX-TrpnX-M_at', 'AFFX-YEL002c/WBP1_at', 'AFFX-YEL018w/_at', 'AFFX-YEL021w/URA3_at', 'AFFX-YEL024w/RIP1_at', 'AFFX-hum_alu_at']\n"
     ]
    }
   ],
   "source": [
    "nan_genes = genes[genes[\"chrom\"].isna()].index.values # genes not associated with chromosomes\n",
    "controls = [gene for gene in nan_genes if gene.startswith(\"AFFX\")]\n",
    "print(controls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1000_at</th>\n",
       "      <th>1001_at</th>\n",
       "      <th>1002_f_at</th>\n",
       "      <th>1003_s_at</th>\n",
       "      <th>1004_at</th>\n",
       "      <th>1005_at</th>\n",
       "      <th>1006_at</th>\n",
       "      <th>1007_s_at</th>\n",
       "      <th>1008_f_at</th>\n",
       "      <th>1009_at</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "      <th>AFFX-YEL002c/WBP1_at</th>\n",
       "      <th>AFFX-YEL018w/_at</th>\n",
       "      <th>AFFX-YEL021w/URA3_at</th>\n",
       "      <th>AFFX-YEL024w/RIP1_at</th>\n",
       "      <th>AFFX-hum_alu_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01_a_D_f_2.CEL</th>\n",
       "      <td>9.521934</td>\n",
       "      <td>7.453767</td>\n",
       "      <td>7.045636</td>\n",
       "      <td>7.743690</td>\n",
       "      <td>7.728274</td>\n",
       "      <td>8.081243</td>\n",
       "      <td>6.927978</td>\n",
       "      <td>9.294152</td>\n",
       "      <td>8.888816</td>\n",
       "      <td>8.290944</td>\n",
       "      <td>...</td>\n",
       "      <td>7.065597</td>\n",
       "      <td>6.803698</td>\n",
       "      <td>6.631766</td>\n",
       "      <td>6.981474</td>\n",
       "      <td>7.003319</td>\n",
       "      <td>6.999630</td>\n",
       "      <td>7.005495</td>\n",
       "      <td>6.864895</td>\n",
       "      <td>7.030665</td>\n",
       "      <td>14.394582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01_a_I_f_2.CEL</th>\n",
       "      <td>10.930684</td>\n",
       "      <td>9.455482</td>\n",
       "      <td>9.233291</td>\n",
       "      <td>9.777128</td>\n",
       "      <td>9.612399</td>\n",
       "      <td>9.845444</td>\n",
       "      <td>9.149639</td>\n",
       "      <td>10.580062</td>\n",
       "      <td>11.505571</td>\n",
       "      <td>9.846817</td>\n",
       "      <td>...</td>\n",
       "      <td>9.366604</td>\n",
       "      <td>9.067484</td>\n",
       "      <td>8.945584</td>\n",
       "      <td>9.118070</td>\n",
       "      <td>9.137492</td>\n",
       "      <td>9.067202</td>\n",
       "      <td>9.129742</td>\n",
       "      <td>9.256544</td>\n",
       "      <td>9.171725</td>\n",
       "      <td>15.494106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01_a_M_f_1.CEL</th>\n",
       "      <td>6.852731</td>\n",
       "      <td>5.298974</td>\n",
       "      <td>5.033266</td>\n",
       "      <td>6.040661</td>\n",
       "      <td>5.890083</td>\n",
       "      <td>5.810144</td>\n",
       "      <td>5.098923</td>\n",
       "      <td>6.240855</td>\n",
       "      <td>7.280948</td>\n",
       "      <td>6.124910</td>\n",
       "      <td>...</td>\n",
       "      <td>5.374457</td>\n",
       "      <td>5.165619</td>\n",
       "      <td>5.032617</td>\n",
       "      <td>5.382904</td>\n",
       "      <td>5.509690</td>\n",
       "      <td>5.113946</td>\n",
       "      <td>5.273243</td>\n",
       "      <td>5.321357</td>\n",
       "      <td>5.321636</td>\n",
       "      <td>12.856782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01_c_D_f_1.CEL</th>\n",
       "      <td>7.285181</td>\n",
       "      <td>6.258114</td>\n",
       "      <td>6.119443</td>\n",
       "      <td>6.631768</td>\n",
       "      <td>6.744592</td>\n",
       "      <td>6.589478</td>\n",
       "      <td>6.156638</td>\n",
       "      <td>7.417750</td>\n",
       "      <td>8.663882</td>\n",
       "      <td>8.203275</td>\n",
       "      <td>...</td>\n",
       "      <td>6.332162</td>\n",
       "      <td>6.278864</td>\n",
       "      <td>6.108778</td>\n",
       "      <td>6.341791</td>\n",
       "      <td>6.339638</td>\n",
       "      <td>6.167335</td>\n",
       "      <td>6.170734</td>\n",
       "      <td>6.734936</td>\n",
       "      <td>6.275118</td>\n",
       "      <td>13.683484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01_c_I_f_2.CEL</th>\n",
       "      <td>11.224543</td>\n",
       "      <td>9.800931</td>\n",
       "      <td>9.407753</td>\n",
       "      <td>10.113212</td>\n",
       "      <td>9.871853</td>\n",
       "      <td>10.495533</td>\n",
       "      <td>9.449701</td>\n",
       "      <td>11.084619</td>\n",
       "      <td>11.969619</td>\n",
       "      <td>10.479006</td>\n",
       "      <td>...</td>\n",
       "      <td>9.443910</td>\n",
       "      <td>9.242560</td>\n",
       "      <td>9.272848</td>\n",
       "      <td>9.435175</td>\n",
       "      <td>9.448927</td>\n",
       "      <td>9.395768</td>\n",
       "      <td>9.306998</td>\n",
       "      <td>9.863770</td>\n",
       "      <td>9.404475</td>\n",
       "      <td>15.494106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_c_I_f_2.CEL</th>\n",
       "      <td>10.834992</td>\n",
       "      <td>9.375171</td>\n",
       "      <td>9.100567</td>\n",
       "      <td>9.766162</td>\n",
       "      <td>9.441591</td>\n",
       "      <td>10.245868</td>\n",
       "      <td>9.065520</td>\n",
       "      <td>10.590598</td>\n",
       "      <td>11.039345</td>\n",
       "      <td>10.225186</td>\n",
       "      <td>...</td>\n",
       "      <td>9.197899</td>\n",
       "      <td>8.932622</td>\n",
       "      <td>8.950766</td>\n",
       "      <td>9.120934</td>\n",
       "      <td>9.018108</td>\n",
       "      <td>9.007320</td>\n",
       "      <td>9.010432</td>\n",
       "      <td>9.093493</td>\n",
       "      <td>8.986246</td>\n",
       "      <td>15.493637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_c_M_f_1.CEL</th>\n",
       "      <td>6.990507</td>\n",
       "      <td>5.671513</td>\n",
       "      <td>5.398236</td>\n",
       "      <td>6.182388</td>\n",
       "      <td>6.088754</td>\n",
       "      <td>6.074909</td>\n",
       "      <td>5.479757</td>\n",
       "      <td>6.328909</td>\n",
       "      <td>7.675669</td>\n",
       "      <td>6.726671</td>\n",
       "      <td>...</td>\n",
       "      <td>5.563038</td>\n",
       "      <td>5.453313</td>\n",
       "      <td>5.228861</td>\n",
       "      <td>5.498717</td>\n",
       "      <td>5.515519</td>\n",
       "      <td>5.379649</td>\n",
       "      <td>5.460068</td>\n",
       "      <td>5.455070</td>\n",
       "      <td>5.488273</td>\n",
       "      <td>12.665336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_d_D_f_2.CEL</th>\n",
       "      <td>9.143998</td>\n",
       "      <td>6.862450</td>\n",
       "      <td>6.357731</td>\n",
       "      <td>7.134995</td>\n",
       "      <td>7.046711</td>\n",
       "      <td>7.730850</td>\n",
       "      <td>6.183967</td>\n",
       "      <td>9.414389</td>\n",
       "      <td>9.065153</td>\n",
       "      <td>8.863422</td>\n",
       "      <td>...</td>\n",
       "      <td>6.258799</td>\n",
       "      <td>6.027364</td>\n",
       "      <td>5.907490</td>\n",
       "      <td>6.171226</td>\n",
       "      <td>6.130500</td>\n",
       "      <td>6.080228</td>\n",
       "      <td>5.966166</td>\n",
       "      <td>6.135691</td>\n",
       "      <td>6.023620</td>\n",
       "      <td>14.244014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_d_I_f_2.CEL</th>\n",
       "      <td>10.963519</td>\n",
       "      <td>9.338467</td>\n",
       "      <td>9.047548</td>\n",
       "      <td>9.696127</td>\n",
       "      <td>9.541697</td>\n",
       "      <td>9.792356</td>\n",
       "      <td>9.073002</td>\n",
       "      <td>10.775091</td>\n",
       "      <td>11.373533</td>\n",
       "      <td>10.350302</td>\n",
       "      <td>...</td>\n",
       "      <td>9.126488</td>\n",
       "      <td>8.891157</td>\n",
       "      <td>8.935299</td>\n",
       "      <td>9.080840</td>\n",
       "      <td>9.166819</td>\n",
       "      <td>9.042159</td>\n",
       "      <td>9.048415</td>\n",
       "      <td>9.007950</td>\n",
       "      <td>9.115568</td>\n",
       "      <td>15.493855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_d_M_f_2.CEL</th>\n",
       "      <td>8.144239</td>\n",
       "      <td>6.028297</td>\n",
       "      <td>5.813491</td>\n",
       "      <td>6.543962</td>\n",
       "      <td>6.395579</td>\n",
       "      <td>6.625704</td>\n",
       "      <td>5.766611</td>\n",
       "      <td>7.342089</td>\n",
       "      <td>8.085729</td>\n",
       "      <td>7.334496</td>\n",
       "      <td>...</td>\n",
       "      <td>5.957419</td>\n",
       "      <td>5.708379</td>\n",
       "      <td>5.659275</td>\n",
       "      <td>5.840375</td>\n",
       "      <td>5.944471</td>\n",
       "      <td>5.770632</td>\n",
       "      <td>5.721384</td>\n",
       "      <td>5.833984</td>\n",
       "      <td>5.880374</td>\n",
       "      <td>12.794010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84 rows × 12600 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  1000_at   1001_at  1002_f_at  1003_s_at   1004_at  \\\n",
       "01_a_D_f_2.CEL   9.521934  7.453767   7.045636   7.743690  7.728274   \n",
       "01_a_I_f_2.CEL  10.930684  9.455482   9.233291   9.777128  9.612399   \n",
       "01_a_M_f_1.CEL   6.852731  5.298974   5.033266   6.040661  5.890083   \n",
       "01_c_D_f_1.CEL   7.285181  6.258114   6.119443   6.631768  6.744592   \n",
       "01_c_I_f_2.CEL  11.224543  9.800931   9.407753  10.113212  9.871853   \n",
       "...                   ...       ...        ...        ...       ...   \n",
       "10_c_I_f_2.CEL  10.834992  9.375171   9.100567   9.766162  9.441591   \n",
       "10_c_M_f_1.CEL   6.990507  5.671513   5.398236   6.182388  6.088754   \n",
       "10_d_D_f_2.CEL   9.143998  6.862450   6.357731   7.134995  7.046711   \n",
       "10_d_I_f_2.CEL  10.963519  9.338467   9.047548   9.696127  9.541697   \n",
       "10_d_M_f_2.CEL   8.144239  6.028297   5.813491   6.543962  6.395579   \n",
       "\n",
       "                  1005_at   1006_at  1007_s_at  1008_f_at    1009_at  ...  \\\n",
       "01_a_D_f_2.CEL   8.081243  6.927978   9.294152   8.888816   8.290944  ...   \n",
       "01_a_I_f_2.CEL   9.845444  9.149639  10.580062  11.505571   9.846817  ...   \n",
       "01_a_M_f_1.CEL   5.810144  5.098923   6.240855   7.280948   6.124910  ...   \n",
       "01_c_D_f_1.CEL   6.589478  6.156638   7.417750   8.663882   8.203275  ...   \n",
       "01_c_I_f_2.CEL  10.495533  9.449701  11.084619  11.969619  10.479006  ...   \n",
       "...                   ...       ...        ...        ...        ...  ...   \n",
       "10_c_I_f_2.CEL  10.245868  9.065520  10.590598  11.039345  10.225186  ...   \n",
       "10_c_M_f_1.CEL   6.074909  5.479757   6.328909   7.675669   6.726671  ...   \n",
       "10_d_D_f_2.CEL   7.730850  6.183967   9.414389   9.065153   8.863422  ...   \n",
       "10_d_I_f_2.CEL   9.792356  9.073002  10.775091  11.373533  10.350302  ...   \n",
       "10_d_M_f_2.CEL   6.625704  5.766611   7.342089   8.085729   7.334496  ...   \n",
       "\n",
       "                AFFX-ThrX-5_at  AFFX-ThrX-M_at  AFFX-TrpnX-3_at  \\\n",
       "01_a_D_f_2.CEL        7.065597        6.803698         6.631766   \n",
       "01_a_I_f_2.CEL        9.366604        9.067484         8.945584   \n",
       "01_a_M_f_1.CEL        5.374457        5.165619         5.032617   \n",
       "01_c_D_f_1.CEL        6.332162        6.278864         6.108778   \n",
       "01_c_I_f_2.CEL        9.443910        9.242560         9.272848   \n",
       "...                        ...             ...              ...   \n",
       "10_c_I_f_2.CEL        9.197899        8.932622         8.950766   \n",
       "10_c_M_f_1.CEL        5.563038        5.453313         5.228861   \n",
       "10_d_D_f_2.CEL        6.258799        6.027364         5.907490   \n",
       "10_d_I_f_2.CEL        9.126488        8.891157         8.935299   \n",
       "10_d_M_f_2.CEL        5.957419        5.708379         5.659275   \n",
       "\n",
       "                AFFX-TrpnX-5_at  AFFX-TrpnX-M_at  AFFX-YEL002c/WBP1_at  \\\n",
       "01_a_D_f_2.CEL         6.981474         7.003319              6.999630   \n",
       "01_a_I_f_2.CEL         9.118070         9.137492              9.067202   \n",
       "01_a_M_f_1.CEL         5.382904         5.509690              5.113946   \n",
       "01_c_D_f_1.CEL         6.341791         6.339638              6.167335   \n",
       "01_c_I_f_2.CEL         9.435175         9.448927              9.395768   \n",
       "...                         ...              ...                   ...   \n",
       "10_c_I_f_2.CEL         9.120934         9.018108              9.007320   \n",
       "10_c_M_f_1.CEL         5.498717         5.515519              5.379649   \n",
       "10_d_D_f_2.CEL         6.171226         6.130500              6.080228   \n",
       "10_d_I_f_2.CEL         9.080840         9.166819              9.042159   \n",
       "10_d_M_f_2.CEL         5.840375         5.944471              5.770632   \n",
       "\n",
       "                AFFX-YEL018w/_at  AFFX-YEL021w/URA3_at  AFFX-YEL024w/RIP1_at  \\\n",
       "01_a_D_f_2.CEL          7.005495              6.864895              7.030665   \n",
       "01_a_I_f_2.CEL          9.129742              9.256544              9.171725   \n",
       "01_a_M_f_1.CEL          5.273243              5.321357              5.321636   \n",
       "01_c_D_f_1.CEL          6.170734              6.734936              6.275118   \n",
       "01_c_I_f_2.CEL          9.306998              9.863770              9.404475   \n",
       "...                          ...                   ...                   ...   \n",
       "10_c_I_f_2.CEL          9.010432              9.093493              8.986246   \n",
       "10_c_M_f_1.CEL          5.460068              5.455070              5.488273   \n",
       "10_d_D_f_2.CEL          5.966166              6.135691              6.023620   \n",
       "10_d_I_f_2.CEL          9.048415              9.007950              9.115568   \n",
       "10_d_M_f_2.CEL          5.721384              5.833984              5.880374   \n",
       "\n",
       "                AFFX-hum_alu_at  \n",
       "01_a_D_f_2.CEL        14.394582  \n",
       "01_a_I_f_2.CEL        15.494106  \n",
       "01_a_M_f_1.CEL        12.856782  \n",
       "01_c_D_f_1.CEL        13.683484  \n",
       "01_c_I_f_2.CEL        15.494106  \n",
       "...                         ...  \n",
       "10_c_I_f_2.CEL        15.493637  \n",
       "10_c_M_f_1.CEL        12.665336  \n",
       "10_d_D_f_2.CEL        14.244014  \n",
       "10_d_I_f_2.CEL        15.493855  \n",
       "10_d_M_f_2.CEL        12.794010  \n",
       "\n",
       "[84 rows x 12600 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sym      NaN\n",
       "chrom    NaN\n",
       "Name: 1155_at, dtype: object"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.loc[\"1155_at\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient</th>\n",
       "      <th>sex</th>\n",
       "      <th>region</th>\n",
       "      <th>lab</th>\n",
       "      <th>chip.version</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rownames</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01_a_D_f_2.CEL</th>\n",
       "      <td>patient_01</td>\n",
       "      <td>female</td>\n",
       "      <td>A.C. cortex</td>\n",
       "      <td>Davis</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01_a_I_f_2.CEL</th>\n",
       "      <td>patient_01</td>\n",
       "      <td>female</td>\n",
       "      <td>A.C. cortex</td>\n",
       "      <td>Irvine</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01_a_M_f_1.CEL</th>\n",
       "      <td>patient_01</td>\n",
       "      <td>female</td>\n",
       "      <td>A.C. cortex</td>\n",
       "      <td>Michigan</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01_c_D_f_1.CEL</th>\n",
       "      <td>patient_01</td>\n",
       "      <td>female</td>\n",
       "      <td>cerebellum</td>\n",
       "      <td>Davis</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01_c_I_f_2.CEL</th>\n",
       "      <td>patient_01</td>\n",
       "      <td>female</td>\n",
       "      <td>cerebellum</td>\n",
       "      <td>Irvine</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_c_I_f_2.CEL</th>\n",
       "      <td>patient_10</td>\n",
       "      <td>female</td>\n",
       "      <td>cerebellum</td>\n",
       "      <td>Irvine</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_c_M_f_1.CEL</th>\n",
       "      <td>patient_10</td>\n",
       "      <td>female</td>\n",
       "      <td>cerebellum</td>\n",
       "      <td>Michigan</td>\n",
       "      <td>v1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_d_D_f_2.CEL</th>\n",
       "      <td>patient_10</td>\n",
       "      <td>female</td>\n",
       "      <td>D.L.P.F. cortex</td>\n",
       "      <td>Davis</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_d_I_f_2.CEL</th>\n",
       "      <td>patient_10</td>\n",
       "      <td>female</td>\n",
       "      <td>D.L.P.F. cortex</td>\n",
       "      <td>Irvine</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_d_M_f_2.CEL</th>\n",
       "      <td>patient_10</td>\n",
       "      <td>female</td>\n",
       "      <td>D.L.P.F. cortex</td>\n",
       "      <td>Michigan</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   patient     sex           region       lab chip.version\n",
       "rownames                                                                  \n",
       "01_a_D_f_2.CEL  patient_01  female      A.C. cortex     Davis           v2\n",
       "01_a_I_f_2.CEL  patient_01  female      A.C. cortex    Irvine           v2\n",
       "01_a_M_f_1.CEL  patient_01  female      A.C. cortex  Michigan           v1\n",
       "01_c_D_f_1.CEL  patient_01  female       cerebellum     Davis           v1\n",
       "01_c_I_f_2.CEL  patient_01  female       cerebellum    Irvine           v2\n",
       "...                    ...     ...              ...       ...          ...\n",
       "10_c_I_f_2.CEL  patient_10  female       cerebellum    Irvine           v2\n",
       "10_c_M_f_1.CEL  patient_10  female       cerebellum  Michigan           v1\n",
       "10_d_D_f_2.CEL  patient_10  female  D.L.P.F. cortex     Davis           v2\n",
       "10_d_I_f_2.CEL  patient_10  female  D.L.P.F. cortex    Irvine           v2\n",
       "10_d_M_f_2.CEL  patient_10  female  D.L.P.F. cortex  Michigan           v2\n",
       "\n",
       "[84 rows x 5 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('bliss')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "97f4452dae6bdba237fd800176b85e526041526e1433dfb6b82689f83ea463b6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
